src/coreclr/jit/targetamd64.h

```diff
@@ -207,32 +207,69 @@
   // Registers no longer containing GC pointers after CORINFO_HELP_ASSIGN_BYREF.
   #define RBM_CALLEE_GCTRASH_WRITEBARRIER_BYREF (RBM_CALLEE_TRASH_NOGC & ~(RBM_RDI | RBM_RSI))
 
-#if 0
-#define REG_VAR_ORDER            REG_EAX,REG_EDX,REG_ECX,REG_ESI,REG_EDI,REG_EBX,REG_ETW_FRAMED_EBP_LIST \
-                                 REG_R8,REG_R9,REG_R10,REG_R11,REG_R14,REG_R15,REG_R12,REG_R13
-#else
-  // TEMPORARY ORDER TO AVOID CALLEE-SAVES
-  // TODO-CQ: Review this and set appropriately
+  // Registers are ordered here to prefer callee trash first, then callee save
```

Comment:
Remind me, but why we prefer callee trash first? They need to be save/restored during the call, right, so in that case, callee save should be better, no?

Reply:
I added a lengthy comment explaining it, but the basics is that we've already been assumed to have used every `callee trash` (aka `volatile` or `caller saved`) register. We only have to save them if the value they contain is `live` across a call boundary, so its not applicable at all to `leaf` methods.

`Callee saved` (aka `non-volatile`) registers on the other hand have an inherent overhead regardless of whether we have calls or not. This is because the method consuming them must preserve the existing value (typically as part of the prologue) and restore that same value before returning (typically as part of the epilogue). They are only beneficial to use if we have a value that needs to live across a call boundary.

I think in a more ideal world LSRA would independently track `callee trash` vs `callee saved` order and it would pick from one of the two pools based on whether the value needs to live across a call boundary. It would do this based on whether a block is hot vs cold and its weight since a `callee trash` register that is spilled once is "equal" to a `callee saved` and so `callee trash` remains "better". But if the value needs to be spilled twice, then `callee saved` becomes cheaper since its just 1 spill/restore needed for the whole method.

Reply:
>  `callee trash`: We only have to save them if the value they contain is live across a call boundary

Right, there is a cost to save/restore them across call boundaries, but `callee save` are cheaper to allocate in my opinion because they are anyway saved and restored during prolog/epilog and hence no separate store/restore needed across call boundaries, and hence should be preferred than `callee trash`.

Related work: https://github.com/dotnet/runtime/pull/76904 and https://github.com/dotnet/runtime/pull/81242

Reply:
> but callee save are cheaper to allocate in my opinion because they are anyway saved and restored during prolog/epilog and hence no separate store/restore needed across call boundaries

Right, they *can* be cheaper in the right circumstances, the inverse is also true:
* A value `x` that is live across 0 calls needs 0 (callee trash) or 1 (callee saved) spills
* A value `x` that is live across 1 call needs 1 (callee trash or callee saved) spill
* A value `x` and value `y` that are live across the same 1 call needs 2 (callee trash or callee saved) spills
* A value `x` that is live across 2 calls needs 2 (callee trash) or 1 (callee saved) spills
* A value `x` that is live across call `A` and a value `y` that is live across call `B` needs 2 (callee trash) or 1 (callee saved) spills

Locally Testing SPMI shows that having the total order as `callee trash, callee saved` vs `callee saved, callee trash` actually makes no difference. This is because of the combination of preferencing done and `interval->preferCalleeSave` tracking we already do which causes `LinearScan::RegisterSelection::select` to already filter down to `callee trash` vs `callee saved` registers. The ordering within each respective group does still matter and is what this PR touches on (this ordering is used to break ties and comes after other filtering/selection is done).

The overall default state for the LSRA is then `callee trash, callee saved` since there are overall more `callee trash` registers, they are frequently preferenced due to their special uses (returns, parameters, required for some instructions, etc), and they are cheaper or the same cost in 3/5 of the scenarios listed above, with those other scenarios already being specially tracked.

Reply:
The `preferCalleeSave` tracking we do isn't ideal and that's partially where #76904 and #81242 come into play.

We also notably don't do tie breaking super efficiently as we currently iterate through all candidates each time comparing against their respective ordering to find the smallest one (when an instruction has no preferences, this is all unused registers which is often around half of the total register count).